pallets/msa/src/lib.rs

```diff
@@ -955,6 +995,94 @@ impl<T: Config> Pallet<T> {
 		}
 		Ok(Some(schemas.0.into()))
 	}
+
+	/// Adds a signature to the PayloadSignatureRegistry based on a virtual "bucket" grouping.
+	/// Check that mortality_block is within bounds. If so, proceed.
+	/// Reset the virtual bucket if necessary, then add the new entry.
+	pub fn register_signature(
+		signature: &MultiSignature,
+		mortality_block: T::BlockNumber,
+	) -> DispatchResult {
+		let current_block = frame_system::Pallet::<T>::block_number();
+
+		if Self::mortality_block_limit(current_block) < mortality_block {
+			Err(Error::<T>::ProofNotYetValid.into())
+		} else if current_block > mortality_block {
+			Err(Error::<T>::ProofHasExpired.into())
+		} else {
+			let bucket_num = Self::bucket_for(mortality_block.into());
+			<PayloadSignatureRegistry<T>>::try_mutate(
+				bucket_num,
+				signature,
+				|maybe_mortality_block| -> DispatchResult {
+					ensure!(maybe_mortality_block.is_none(), Error::<T>::SignatureAlreadySubmitted);
+
+					Self::reset_virtual_bucket_if_needed(mortality_block, bucket_num)?;
+					*maybe_mortality_block = Some(mortality_block);
+					Ok(())
+				},
+			)
+		}
+	}
+
+	// Check if enough blocks have passed to reset bucket mortality storage.
+	// If so:
+	//     1. delete all the stored bucket/signature alues with key1 = bucket num
+	//     2. set the bucket's mortality block to the new value
+	// If not, don't do anything.
+	fn reset_virtual_bucket_if_needed(
+		mortality_block: T::BlockNumber,
+		bucket_num: T::BlockNumber,
+	) -> DispatchResult {
+		<MortalityBlockOf<T>>::try_mutate(bucket_num, |maybe_mortality_block| -> DispatchResult {
+			let bucket_mortality_block = Self::bucket_mortality_block(mortality_block);
+
+			match maybe_mortality_block {
+				None => Err(Error::<T>::NoSuchBucket.into()),
+				Some(mortality_block_for_bucket) => {
+					if mortality_block > *mortality_block_for_bucket {
+						Self::delete_signatures_for_bucket(bucket_num)?;
+						*maybe_mortality_block = Some(bucket_mortality_block);
+					} else if *mortality_block_for_bucket == T::BlockNumber::default() {
+						*maybe_mortality_block = Some(bucket_mortality_block);
```

Comment:
maybe_mortality_block can be out of if ,they look the same Some(bucket_mortality_block) only a delete step in taken when conditions meet

Reply:
@shannonwells  I have noticed when it is the first time write in MortalityBlockOf while registering msa or adding provider etc. it will always have ```None``` for ```maybe_mortality_block```  for a given ```bucket_num```